src/builder.rs

```diff
@@ -794,25 +794,25 @@ impl Builder {
     }
 
     // SubType: <I>(&self, op, lhs: &IntValue<I>, rhs: &IntValue<I>, name) -> IntValue<bool> { ?
-    pub fn build_int_compare(&self, op: IntPredicate, lhs: &IntValue, rhs: &IntValue, name: &str) -> IntValue {
+    pub fn build_int_compare<T: IntMathValue>(&self, op: IntPredicate, lhs: &T, rhs: &T, name: &str) -> T {
         let c_string = CString::new(name).expect("Conversion to CString failed unexpectedly");
 
         let value = unsafe {
             LLVMBuildICmp(self.builder, op.as_llvm_predicate(), lhs.as_value_ref(), rhs.as_value_ref(), c_string.as_ptr())
         };
 
-        IntValue::new(value)
+        T::new(value)
     }
 
     // SubType: <F>(&self, op, lhs: &FloatValue<F>, rhs: &FloatValue<F>, name) -> IntValue<bool> { ?
-    pub fn build_float_compare(&self, op: FloatPredicate, lhs: &FloatValue, rhs: &FloatValue, name: &str) -> IntValue {
+    pub fn build_float_compare<T: FloatMathValue>(&self, op: FloatPredicate, lhs: &T, rhs: &T, name: &str) -> <<T::BaseType as FloatMathType>::ConvType as IntMathType>::ValueType {
```

Comment:
I think the issue with `build_int_compare` and `build_float_compare` returning `T` and `<<T::BaseType as FloatMathType>::ConvType as IntMathType>::ValueType` (man is that ugly) respectively is that the return type is actually `IntValue<bool>` or `VecValue<IntValue<bool>>`.

This will be rendered incorrect once subtypes are implemented. IE input of `IntValue<32bit>` will return `IntValue<32bit>` when it should be `IntValue<bool>`

I think the way we might need to work around this with subtypes is to add more associated types to go from `IntValue<X>` -> `IntValue<bool>` & `VecValue<IntValue<X>>` -> `Vec<IntValue<bool>>` (and likewise for floats) somehow

...so I think this is fine as is, unless you can think of a good way to do that here. I would greatly appreciate adding another comment under the subtype comment either describing the above or pasting the URL to this comment

Reply:
Yeah, there's possibly a similar problem with functions like `build_int_cast`, if the `Type` types are also going to have subtypes.

There's currently [an RFC being worked on](https://github.com/rust-lang/rust/issues/44265) for generic associated types, which _I think_ would allow us to do something like `<<T::BaseType as FloatMathType>::ConvType<bool> as IntMathType>::ValueType` (notice how `ConvType` has a generic parameter with the type of integer to 'return') - a similar thing would also work for `BaseType` for the cast functions. But yeah, it's pretty ugly.